Funnel tests through the command line

cowboyd opened the pull request:
We want to ship a command line along with this ruby library.

This change makes the test cases flow throw the CLI so that we get to

a) test that the cli is working
b) consume the client api to see if it's good or not.

In terms of validating which data is necessary for each request, all of the inputs for the client have been concentrated into a `Settings` object. The idea being that if any code _asks_ for a setting that isn't there, that's when it should blow up. Laziness FTW.  For example, if through fulfilling a request, you need to ask the settings for the token, then by definition the token must be set. On the other hand, if you _don't_ need a token, then the code will never actually validate that you have one.

cowboyd committed:
Rig up with test harness

Funnel all of the tests through the command line.

wwilsman commented on fixtures/vcr_cassettes/okapi-cli-specs.yml:
```diff
@@ -0,0 +1,72 @@
+        } ]
+    http_version: 
+  recorded_at: Thu, 05 Oct 2017 21:53:45 GMT
+recorded_with: VCR 3.0.3
```
This uses the old okapi-1.5 cluster. We should probably use `https://okapi2.frontside.io` (or switch the new okapi-2.x cluster to `https://okapi.frontside.io`).

Particularly, _most_ internal endpoints will now complain with
```
No suitable module found for http://okapi2.frontside.io/_/proxy/modules
```
if you try to hit the endpoint with the `X-Okapi-Tenant` header set (unless it's set to `okapi.supertanent`)

cowboyd replied:
I'm totally ok using either one for the tests.

> if you try to hit the endpoint with the X-Okapi-Tenant header set (unless it's set to okapi.supertanent)

"with" or "without"

wwilsman replied:
With

![image](https://user-images.githubusercontent.com/5005153/31286359-87c6fd4c-aa83-11e7-92b8-f8fb890b7c13.png)

wwilsman replied:
Without the header is fine, but we have to make sure we don't add the tenant to these internal endpoints

![image](https://user-images.githubusercontent.com/5005153/31286422-b38cf49a-aa83-11e7-90bf-7425b99b3ed4.png)

wwilsman replied:
`okapi.supertenant` also works, so maybe instead of removing the tenant for those endpoints, we ensure that the tenant is the supertenant for those endpoints.

cowboyd replied:
I see, this is actually related to the change around settings too. It should "just work" In other words, i made the change to _not_ actually use either the tenant or the token unless you explicity ask for it. even if the `--tenant` or `OKAPI_TENANT` options are set.

```ruby
client.get('/_/proxy/modules') # does not set tenant header
client.tenant.get('/_/proxy/modules') # does set tenant header.
```

wwilsman replied:
~I see. But `client.has_interface?('configuration')` will fail because there is no tenant for `/_/proxy/tenants/#{@settings.tenant}/interfaces/#{interface_name}`.~

~However `client.tenant.has_interface?('configuration')` will also fail because the request automatically sets the `x-okapi-tenant` header.~

Aha, `@settings.tenant` comes from the environment and `client.tenant` is what sets the header.

cowboyd committed:
Empty check environment variables

*speculatively*, on Linux non-present environment variables are just
 empty strings and not `nil`. This abstracts instance and environment
 access for every setting so that it can check to make sure that the
 environment variable is actually present.
